compatibility/bazel_tools/create-daml-app/Main.hs

```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
+          -- we need 'npm-cli.js' in the path for the following test
+          mbOldPath <- getEnv "PATH"
+          setEnv "PATH" (takeDirectory npmPath <> (searchPathSeparator : fromMaybe "" mbOldPath)) True
+          callProcess npmPath ["run", "test", "--ci", "--all", "--scripts-prepend-node-path"]
 
 addTestDependencies :: FilePath -> FilePath -> IO ()
 addTestDependencies packageJsonFile extraDepsFile = do
     packageJson <- readJsonFile packageJsonFile
     extraDeps <- readJsonFile extraDepsFile
     let newPackageJson = Aeson.lodashMerge packageJson extraDeps
-    Aeson.encodeFile packageJsonFile newPackageJson
-  where
-    readJsonFile :: FilePath -> IO Aeson.Value
-    readJsonFile path = do
-        -- Read file strictly to avoid lock being held when we subsequently write to it.
-        mbContent <- Aeson.decodeFileStrict' path
-        case mbContent of
-            Nothing -> fail ("Could not decode JSON object from " <> path)
-            Just val -> return val
+    BSL.writeFile packageJsonFile (Aeson.encode newPackageJson)
+
+readJsonFile :: FilePath -> IO Aeson.Value
+readJsonFile path = do
+    -- Read file strictly to avoid lock being held when we subsequently write to it.
+    content <- BSL.fromStrict <$> BS.readFile path
+    case Aeson.decode content of
+        Nothing -> error ("Could not decode JSON object from " <> path)
+        Just val -> return val
+
+extractTarGz :: FilePath -> FilePath -> IO ()
+extractTarGz targz outDir = do
+    runConduitRes
+        $ Conduit.sourceFile targz
+        .| Zlib.ungzip
+        .| Tar.untar (restoreFile (\a b -> fail (T.unpack $ a <> b)) outDir)
+
+setupNpmEnv :: FilePath -> [(TsLibrary, FilePath)] -> IO ()
+setupNpmEnv uiDir libs = do
+  forM_ libs $ \(tsLib, path) -> do
+    let name = tsLibraryName tsLib
+    let uiLibPath = uiDir </> name
+    extractTarGz path uiLibPath
+    patchTsDependencies uiDir $ uiLibPath </> "package.json"
+
+-- | Overwrite dependencies to our TypeScript libraries to point to local file dependencies in the
+-- 'ui' directory in the specified package.json file.
+patchTsDependencies :: FilePath -> FilePath -> IO ()
+patchTsDependencies uiDir packageJsonFile = do
+  packageJson0 <- readJsonFile packageJsonFile
+  case packageJson0 of
+    Aeson.Object packageJson ->
+      case HMS.lookup "dependencies" packageJson of
+        Just (Aeson.Object dependencies) -> do
+          let depNames = HMS.keys dependencies
+          -- patch dependencies to point to local files if they are present in the package.json
+          let patchedDeps =
+                HMS.fromList
+                  ([ ( "@daml.js/create-daml-app"
+                     , Aeson.String $
+                       T.pack $
+                       "file:" <> "./daml.js/create-daml-app-0.1.0")
+                   | "@daml.js/create-daml-app" `elem` depNames
+                   ] ++
+                   [ (depName, Aeson.String $ T.pack $ "file:" <> libRelPath)
```

Comment:
technically relative file urls are invalid. I guess npm handles them given that this test passes but perhaps worth keeping in mind.

Reply:
I didn't know that! We did have absolute file path in there first and changed it to relative ones.

Reply:
"Technically" as in the RFC for URLs? NPM seems to be fine with them [per their documentation of package.json](https://docs.npmjs.com/files/package.json#local-paths).

Reply:
yep, according to the rfc you need a leading slash.